File: js-ts/matrices.js
/**
 * Matrices en JavaScript - Estructura de Datos, Unidad 1.
 *
 * JavaScript no tiene un tipo matriz nativo: una matriz se representa como un
 * arreglo de arreglos, donde cada arreglo interno es una fila.
 *
 * Ejecución:
 *   node js-ts/matrices.js
 */

const FILAS = 3;
const COLUMNAS = 3;

/** Representa la matriz como [[a, b], [c, d]] (igual que print en Python). */
function formatear(matriz) {
  return `[${matriz.map((fila) => `[${fila.join(", ")}]`).join(", ")}]`;
}

// 1. Declaración e inicialización
/** Crea una matriz filas x columnas con valores consecutivos desde 1. */
function crearMatriz(filas = FILAS, columnas = COLUMNAS) {
  // Declaración: cada fila debe ser un arreglo distinto. Escribir
  // new Array(filas).fill(new Array(columnas).fill(0)) crearía filas que
  // son EL MISMO arreglo.
  const matriz = Array.from({ length: filas }, () => new Array(columnas).fill(0));
  // Inicialización: valores del 1 al filas*columnas (1..9 en una 3x3).
  // Equivalente literal: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
  let valor = 1;
  for (let i = 0; i < filas; i++) {
    for (let j = 0; j < columnas; j++) {
      matriz[i][j] = valor;
      valor++;
    }
  }
  return matriz;
}

function main() {
  console.log("=== MATRICES EN JAVASCRIPT ===");

  console.log("\n1. Declaración e inicialización");
  const matriz = crearMatriz();
  console.log(`  Matriz ${matriz.length}x${matriz[0].length} con valores del 1 al 9: ${formatear(matriz)}`);
}

main();

File: python/matrices.py
"""
Matrices en Python - Estructura de Datos, Unidad 1.

Python no tiene un tipo matriz nativo: una matriz se representa como una
lista de listas, donde cada lista interna es una fila.

Ejecución:
    python python/matrices.py
"""

FILAS = 3
COLUMNAS = 3


# 1. Declaración e inicialización
def crear_matriz(filas: int = FILAS, columnas: int = COLUMNAS) -> list[list[int]]:
    """Crea una matriz filas x columnas con valores consecutivos desde 1."""
    # Declaración: cada fila debe ser una lista distinta. Escribir
    # [[0] * columnas] * filas crearía filas que son LA MISMA lista.
    matriz = [[0] * columnas for _ in range(filas)]
    # Inicialización: valores del 1 al filas*columnas (1..9 en una 3x3).
    # Equivalente literal: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    valor = 1
    for i in range(filas):
        for j in range(columnas):
            matriz[i][j] = valor
            valor += 1
    return matriz


def main() -> None:
    print("=== MATRICES EN PYTHON ===")

    print("\n1. Declaración e inicialización")
    matriz = crear_matriz()
    print(f"  Matriz {len(matriz)}x{len(matriz[0])} con valores del 1 al 9: {matriz}")


if __name__ == "__main__":
    main()
